Gateway/Request/BasicParameter/CurrencyDataBuilder.php

```diff
@@ -0,0 +1,124 @@
+use Magento\Payment\Gateway\Request\BuilderInterface;
+use Magento\Sales\Model\Order;
+
+class CurrencyDataBuilder implements BuilderInterface
+{
+    private const CURRENCY = 'currency';
+
+    /** @var Factory */
+    private $configProviderMethodFactory;
+
+    /**
+     * @var string
+     */
+    private string $currency;
+
+    /**
+     * @var array
+     */
+    private array $allowedCurrencies;
+
+    /**
+     * Constructor
+     *
+     * @param Factory $configProviderMethodFactory
+     */
+    public function __construct(
+        Factory $configProviderMethodFactory
+    ) {
+        $this->configProviderMethodFactory = $configProviderMethodFactory;
+    }
+
+    /**
+     * @inheritdoc
+     * @throws Exception
+     */
+    public function build(array $buildSubject): array
+    {
+        $paymentDO = SubjectReader::readPayment($buildSubject);
+        $order = $paymentDO->getOrder()->getOrder();
+        $this->setAllowedCurrencies($paymentDO->getPayment()->getMethodInstance());
+
+        return [
+            self::CURRENCY => $this->getCurrency($order)
+        ];
+    }
+
+    /**
+     * Get Currency
+     *
+     * @param Order|null $order
+     * @return string
+     * @throws Exception
+     */
+    public function getCurrency($order = null)
+    {
+        if (empty($this->currency)) {
+            $this->setCurrency($order);
+        }
+
+        return $this->currency;
+    }
+
+    /**
+     * Set Currency
+     *
+     * @param Order $order
+     * @return $this
+     * @throws Exception
+     */
+    public function setCurrency($order)
+    {
+        $allowedCurrencies = $this->getAllowedCurrencies();
+        if (in_array($order->getOrderCurrencyCode(), $allowedCurrencies)) {
+            $this->currency = $order->getOrderCurrencyCode();
+        } elseif (in_array($order->getBaseCurrencyCode(), $allowedCurrencies)) {
+            $this->currency = $order->getBaseCurrencyCode();
+        } else {
+            throw new \Buckaroo\Magento2\Exception(
+                __("The selected payment method does not support the selected currency or the store's base currency.")
+            );
+        }
+
+        return $this;
+    }
+
+    /**
+     * Get Allowed Currencies
+     *
+     * @param \Buckaroo\Magento2\Model\Method\BuckarooAdapter $methodInstance
+     * @return array
+     * @throws Exception
+     */
+    public function getAllowedCurrencies($methodInstance = null)
+    {
+        if (empty($this->allowedCurrencies)) {
+            $this->setAllowedCurrencies($methodInstance);
+        }
+
+        return $this->allowedCurrencies;
+    }
+
+    /**
+     * Set Allowed Currencies
+     *
+     * @param \Buckaroo\Magento2\Model\Method\BuckarooAdapter $methodInstance
+     * @return $this
+     * @throws \Buckaroo\Magento2\Exception
+     */
+    private function setAllowedCurrencies($methodInstance)
+    {
+        $method = $methodInstance->getCode() ?? 'buckaroo_magento2_ideal';
```

Comment:
why set here the default as `buckaroo_magento2_ideal`  and not throw a exception?

Reply:
app/code/Buckaroo/Magento2/Gateway/Http/TransactionBuilder/Order.php:353 I copied the logic from here, this file was modified in commit: BP-391 Add payment method PayPerEmail (Create Paylink only)  9/23/2020 11:42 AM